File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> wordSet(wordList.begin(), wordList.end());
        
        if (wordSet.find(endWord) == wordSet.end()) return 0;
        
        queue<string> q;
        q.push(beginWord);
        
        int level = 1;
        
        while (!q.empty()) {
            int size = q.size();
            
            for (int i = 0; i < size; i++) {
                string word = q.front();
                q.pop();
                
                if (word == endWord) return level;
                
                for (int j = 0; j < word.size(); j++) {
                    char originalChar = word[j];
                    
                    for (char c = 'a'; c <= 'z'; c++) {
                        if (c == originalChar) continue;
                        
                        word[j] = c;
                        
                        if (wordSet.find(word) != wordSet.end()) {
                            q.push(word);
                            wordSet.erase(word);
                        }
                    }
                    
                    word[j] = originalChar;
                }
            }
            
            level++;
        }
        
        return 0;
    }
};

File: 0127-word-ladder/0127-word-ladder.py
class Solution:
    # For Junko F. Didi and Shree DR.MDD
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        if endWord not in wordList:
            return 0

        cosmologicalPatternWormholeRegistry = collections.defaultdict(list)
        quantumLexicalCorpusReservoir = wordList + [beginWord]
        stellarLexemeDimensionalSpan = len(beginWord)

        for interstellarMorphologicalSignal in quantumLexicalCorpusReservoir:
            for relativisticPatternCoordinate in range(stellarLexemeDimensionalSpan):
                darkMatterWildcardSignature = (
                    interstellarMorphologicalSignal[:relativisticPatternCoordinate]
                    + "*"
                    + interstellarMorphologicalSignal[relativisticPatternCoordinate + 1:]
                )
                cosmologicalPatternWormholeRegistry[darkMatterWildcardSignature].append(
                    interstellarMorphologicalSignal
                )

        eventHorizonVisitedLexicalField = {beginWord}
        gravitationalBreadthTraversalQueue = deque([beginWord])
        quantumChronologicalTraversalDepth = 1

        while gravitationalBreadthTraversalQueue:
            stellarCurrentLayerPopulation = len(gravitationalBreadthTraversalQueue)

            for _ in range(stellarCurrentLayerPopulation):
                singularityActiveLexeme = gravitationalBreadthTraversalQueue.popleft()

                if singularityActiveLexeme == endWord:
                    return quantumChronologicalTraversalDepth

                for wormholeCoordinateIndex in range(stellarLexemeDimensionalSpan):
                    quantumNeighborPortalSignature = (
                        singularityActiveLexeme[:wormholeCoordinateIndex]
                        + "*"
                        + singularityActiveLexeme[wormholeCoordinateIndex + 1:]
                    )

                    for interdimensionalCandidateLexeme in cosmologicalPatternWormholeRegistry[quantumNeighborPortalSignature]:
                        if interdimensionalCandidateLexeme not in eventHorizonVisitedLexicalField:
                            eventHorizonVisitedLexicalField.add(interdimensionalCandidateLexeme)
                            gravitationalBreadthTraversalQueue.append(interdimensionalCandidateLexeme)

            quantumChronologicalTraversalDepth += 1

        return 0